### DISCUSSION TITLE
IDL @optional fields

### DISCUSSION THREAD

**Discussion by stephanhuber15** [OpenDDS/OpenDDS#2536] [Q&A] [answered]
Hi, I am new to OpenDDS, but have been working my way through the developer's guide. I am interfacing with an existing DDS that I was provided the IDL for. RTPS is working and I am triggering DataReaderListerners for some topics (on_data_available()), but not all. I see an issue in my openDDS log for one of the topics "...State_TDataReaderImpl::ddsdemarshal deserialization failed, dropping sample." That particular topic has @optional fields, which I assume are not being populated and causing issues. 

It does look like the @optional feature is currently unsupported according to 16.6.1. Unfortunately, I do not have control over the IDL is there anyway to dump the raw data out and try to make sense of it myself?

**Comment by mitza-oci** [comment, marked as answer]
Could you share the IDL or a redacted version of it?  There may be some things we can suggest based on use of XCDR1/XCDR2 mutable/final/appendable, etc.

**Reply by stephanhuber15** [reply]
Adam, thanks for the quick reply. I will attach a heavily redacted/modified version that only has the pertinent structs. It won't let me upload as an IDL, so extension was changed to txt
[service.txt](https://github.com/objectcomputing/OpenDDS/files/6245888/service.txt)

**Reply by mitza-oci** [reply]
If you are not passing the `--default-extensibility` option to `opendds_idl`, then these are all Appendable.
Do you know if the DataWriter is encoding in XCDR1 or 2?

Which IDL-to-language mapping(s) do you need to support?  Options are:
- C++11 mapping
- C++ classic mapping
- Java

**Reply by stephanhuber15** [reply]
--default-extensibiilty is not being passed into opendds_idl. So I should remove the @optional annotations for each field and add @appendable one line before the struct:

@appendable
struct State_T
{
....

I am using C++ classic mapping.

The documentation states conformance to xtypes v1.3, so I would assume XCDR2? Can I get that information from datawriter?

**Reply by mitza-oci** [reply]
> --default-extensibility is not being passed into opendds_idl. So I should remove the @optional annotations for each field and add @appendable one line before the struct:

No, I'm not recommending any changes.  I'm just asking questions to learn about your use case.
 
> I am using C++ classic mapping.

If support for `@optional` was only in the C++11 mapping, would you switch to it?

> The documentation states conformance to xtypes v1.3, so I would assume XCDR2? Can I get that information from datawriter?

An OpenDDS DataReader can, with some limitations, receive data from an XCDR1 DataWriter.  The DataWriter's use of XCDR1/2 may be based on QoS (DataRepresentationQosPolicy) or inherent capabilities or configuration of that non-OpenDDS system.

**Reply by stephanhuber15** [reply]
Ok, yes I would switch to C++11 mapping if there was support for @optional.
